doc/source/reference/arrays.promotion.rst

```diff
@@ -0,0 +1,263 @@
+the rules:
+* NumPy comparisons with Python integers or mixed precision integers always
+  return the correct result.  The inputs will never be cast in a way which
+  loses precision.
+* Equality comparisons between types which cannot be promoted will be
+  considered all ``False`` (equality) or all ``True`` (not-equal).
+* Unary math functions like ``np.sin`` that always return floating point
+  values, accept any Python integer input by converting it to ``float64``.
+* Division always returns floating point values and thus also allows divisions
+  between any NumPy integer with any Python integer value by casting both
+  to ``float64``.
+
+It should be noted that while these exceptions can apply to more functions,
+NumPy may choose to not implement them in all such cases.
+
+Promotion of non-numerical datatypes
+------------------------------------
+
+NumPy extends the promotion to non-numerical types, although in most cases
+promotion is not well defined and simply rejected.
+
+The following rules apply:
+* NumPy byte strings (``np.bytes_``) can be promoted to unicode strings
+  (``np.str_``).  A ``result_type``
+* For some purposes NumPy will promote almost any other datatype to strings.
+  This applies to array creation or concatenation.
+* The array constructers like ``np.array()`` will use ``object`` dtype when
+  there is no viable promotion.
+* Structured dtypes can promote when their field names and order matches.
+  In that case all fields are promoted individually.
+* NumPy ``timedelta`` can in some cases promote with integers.
+
+.. note::
+    Some of these rules describe the state as of NumPy 2.0 and details
+    may be surprising and are under consideration to be change in the future.
+    However, changes always have to be weighed against backwards compatibility
+    concerns.
+
+Details of promoted ``dtype`` instances
+---------------------------------------
+The above promotion mainly refers to the behavior when mixing different DType
+classes.
+A ``dtype`` instance attached to an array can carry additional information
+such as byte-order, metadata, string length, or exact structured dtype layout.
+
+While the string length or field names of a structured dtype are important,
+NumPy considers byte-order, metadata, and the exact layout of a structured
+dtype as storage details.
+During promotion NumPy does *not* take these storage details into account:
+* Byte-order is converted to native byte-order.
+* Metadata attached to the dtype may or may not be preserved.
+* Resulting structured dtypes will be packed (but aligned if inputs were).
+
+This behaviors is the best behavior for most programs where storage details
+are not relevant to the final results, but the use of incorrect byte-order
+could drastically slow down evaluation.
+
+
+.. [#hist-reasons]: To a large degree, this may just be for choices made early
+   on in NumPy's predecessors even.  You may find some more details also in
```

Comment:
Second sentence: For more details, see `NEP...`

Reply:
adopted